clusterstate/provider/azure.go

```diff
@@ -0,0 +1,192 @@
+	if config.AzureClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.AzureClient must not be empty", config)
+	}
+	if config.GuestFramework == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.GuestFramework must not be empty", config)
+	}
+	if config.HostFramework == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.HostFramework must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	if config.ClusterID == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.ClusterID must not be empty", config)
+	}
+
+	a := &Azure{
+		azureClient:    config.AzureClient,
+		guestFramework: config.GuestFramework,
+		hostFramework:  config.HostFramework,
+		logger:         config.Logger,
+
+		clusterID: config.ClusterID,
+	}
+
+	return a, nil
+}
+
+func (a *Azure) InstallTestApp() error {
+	var err error
+
+	var apprClient *apprclient.Client
+	{
+		c := apprclient.Config{
+			Fs:     afero.NewOsFs(),
+			Logger: a.logger,
+
+			Address:      CNRAddress,
+			Organization: CNROrganization,
+		}
+
+		apprClient, err = apprclient.New(c)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	var helmClient *helmclient.Client
+	{
+		c := helmclient.Config{
+			Logger:    a.logger,
+			K8sClient: a.guestFramework.K8sClient(),
+
+			RestConfig: a.guestFramework.RestConfig(),
+		}
+
+		helmClient, err = helmclient.New(c)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		err = helmClient.EnsureTillerInstalled()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	// Install the e2e app chart in the guest cluster.
+	{
+		a.logger.Log("level", "debug", "message", "installing e2e-app for testing")
+
+		tarballPath, err := apprClient.PullChartTarball(ChartName, ChartChannel)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		err = helmClient.InstallFromTarball(tarballPath, ChartNamespace)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	return nil
+}
+
+func (a *Azure) RebootMaster() error {
+	resourceGroupName := a.clusterID
+	masterVMName := fmt.Sprintf("%s-Master-1", a.clusterID)
+	_, err := a.azureClient.VirtualMachineClient.Restart(context.TODO(), resourceGroupName, masterVMName)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (a *Azure) ReplaceMaster() error {
+	customObject, err := a.hostFramework.G8sClient().ProviderV1alpha1().AzureConfigs("default").Get(a.clusterID, metav1.GetOptions{})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	// Change virtual machine size to trigger replacement of existing master node.
+	customObject.Spec.Azure.Masters[0].VMSize = VirtualMachineSize
+
+	_, err = a.hostFramework.G8sClient().ProviderV1alpha1().AzureConfigs("default").Update(customObject)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (a *Azure) WaitForAPIDown() error {
```

Comment:
I think we could use composition here, this code is the same for the other provider. We could have a separate type with the common code, and include it as an anonymous field in each provider type, WDYT?

Reply:
Yes it would be nice to remove the duplicate code.

Reply:
This was sorted by the restructuring.